### PR TITLE
Implement BindingMode for pattern matching.

### LINKED ISSUE #888: Implement match ergonomics (binding modes) in type inference
E.g.
```rust
let x = Some(1);
match &x {
    Some(n) => n, // n should be &i32
}
match &mut x {
    Some(n) => n, // n should be &mut i32
}
```
and so on. We currently just infer wrong types in this situation. To fix this, when walking through patterns with `infer_pat`, we need to keep track of the 'binding mode' and apply it when creating bindings. This is also explained in the [rust reference](https://doc.rust-lang.org/reference/patterns.html#binding-modes); and I think rustc handles it [here](https://github.com/rust-lang/rust/blob/e17c48e2f21eefd59748e364234efc7037a3ec96/src/librustc_typeck/check/_match.rs#L77).

### PR DESCRIPTION
**Pull request by mjkillough**
Implement `BindingMode` for pattern matching, so that types can be
correctly inferred using match ergonomics. The binding mode defaults to
`Move` (referred to as 'BindingMode::BindByValue` in rustc), and is
updated by automatic dereferencing of the value being matched.

Fixes #888.

 - [Binding modes in The Reference](https://doc.rust-lang.org/reference/patterns.html#binding-modes)
 - [`rustc` implementation](https://github.com/rust-lang/rust/blob/e17c48e2f21eefd59748e364234efc7037a3ec96/src/librustc_typeck/check/_match.rs#L77) (and [definition of `BindingMode`](https://github.com/rust-lang/rust/blob/e957ed9d10ec589bdd523b88b4b44c41b1ecf763/src/librustc/ty/binding.rs))
 - [Match Ergonomics RFC](https://github.com/rust-lang/rfcs/blob/master/text/2005-match-ergonomics.md#binding-mode-rules)

### CODE REVIEW

File: crates/ra_hir/src/ty/tests.rs [hunk 1]
```diff
@@ -772,60 +774,89 @@ fn test(x: &i32) {
 
     let lambda = |a: u64, b, c: i32| { a + b; c };
 
+    let A(n) = &A(1);
+    let A(n) = &mut A(1);
+
+    let v = &(1, &2);
+    let (_, &w) = v;
```
Comment (mjkillough) [inline crates/ra_hir/src/ty/tests.rs]: This tests the [`else if let Pat::Ref { .. }`](https://github.com/rust-analyzer/rust-analyzer/pull/982/files#diff-680da4262b0a53439ea60a27e00f9494R615), which is rust-lang/rust#46688.

File: crates/ra_hir/src/ty/infer.rs [hunk 2]
```diff
@@ -63,6 +63,24 @@ enum ExprOrPatId {
 
 impl_froms!(ExprOrPatId: ExprId, PatId);
 
+/// Binding modes inferred for patterns.
+/// https://doc.rust-lang.org/reference/patterns.html#binding-modes
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+enum BindingMode {
```
Comment (mjkillough) [inline crates/ra_hir/src/ty/infer.rs]: I wasn't sure where to put this new `enum`. As it is only used by inference, I left it here for now.

File: crates/ra_hir/src/ty/infer.rs [hunk 3]
```diff
@@ -63,6 +63,24 @@ enum ExprOrPatId {
...
+    Move,
```
Comment (mjkillough) [inline crates/ra_hir/src/ty/infer.rs]: `rustc` calls this `BindByValue` (and takes a mutability argument).

I chose to call it `Move` to match the reference documentation, and the match ergonomics RFC which refer to it that way. I'm very happy to change it if consistency with `rustc` is more important.

File: crates/ra_hir/src/ty/infer.rs [hunk 4]
```diff
@@ -559,15 +584,42 @@ impl<'a, D: HirDatabase> InferenceContext<'a, D> {
...
         let body = Arc::clone(&self.body); // avoid borrow checker problem
 
+        let is_non_ref_pat = match &body[pat] {
+            Pat::Tuple(..)
+            | Pat::TupleStruct { .. }
+            | Pat::Struct { .. }
+            | Pat::Range { .. }
+            | Pat::Slice { .. } => true,
+            // TODO: Path/Lit might actually evaluate to ref, but inference is unimplemented.
+            Pat::Path(..) | Pat::Lit(..) => true,
```
Comment (mjkillough) [inline crates/ra_hir/src/ty/infer.rs]: [These are the relevant bits](https://github.com/rust-lang/rust/blob/e17c48e2f21eefd59748e364234efc7037a3ec96/src/librustc_typeck/check/_match.rs#L59-L72) of the `rustc` implementation. Though I am not sure we're able to implement those here yet?

File: crates/ra_hir/src/ty/infer.rs [hunk 5]
```diff
@@ -559,15 +584,42 @@ impl<'a, D: HirDatabase> InferenceContext<'a, D> {
...
+                    BindingMode::Move => BindingMode::Ref(*mutability),
+                    BindingMode::Ref(Mutability::Shared) => BindingMode::Ref(Mutability::Shared),
+                    BindingMode::Ref(Mutability::Mut) => BindingMode::Ref(*mutability),
+                }
+            }
+        } else if let Pat::Ref { .. } = &body[pat] {
+            default_bm = BindingMode::Move;
+        }
+
+        // Lose mutability.
```
Comment (matklad) [inline crates/ra_hir/src/ty/infer.rs]: 👍

File: crates/ra_hir/src/ty/infer.rs [hunk 6]
```diff
@@ -700,7 +761,7 @@ impl<'a, D: HirDatabase> InferenceContext<'a, D> {
             }
             Expr::For { iterable, body, pat } => {
                 let _iterable_ty = self.infer_expr(*iterable, &Expectation::none());
-                self.infer_pat(*pat, &Ty::Unknown);
+                self.infer_pat(*pat, &Ty::Unknown, BindingMode::Move);
```
Comment (matklad) [inline crates/ra_hir/src/ty/infer.rs]: Micro nit, but this seems like a good place for Default ?

File: crates/ra_hir/src/ty/infer.rs [hunk 7]
```diff
@@ -559,15 +584,42 @@ impl<'a, D: HirDatabase> InferenceContext<'a, D> {
...
+            while let Ty::Ref(inner, mutability) = expected {
+                expected = inner;
+                default_bm = match default_bm {
+                    BindingMode::Move => BindingMode::Ref(*mutability),
+                    BindingMode::Ref(Mutability::Shared) => BindingMode::Ref(Mutability::Shared),
+                    BindingMode::Ref(Mutability::Mut) => BindingMode::Ref(*mutability),
+                }
+            }
+        } else if let Pat::Ref { .. } = &body[pat] {
+            default_bm = BindingMode::Move;
```
Comment (flodiebold) [inline crates/ra_hir/src/ty/infer.rs]: a similar comment as in rustc might be appropriate here?

File: crates/ra_hir/src/ty/infer.rs (hunk 4 above)
Comment (flodiebold) [reply crates/ra_hir/src/ty/infer.rs]: hm. somehow I was sure we supported literal patterns, but apparently we don't yet :smile: those should be easy to add, though. Path patterns to consts are supported. We'll probably need to extract the 'resolving' part from `infer_path_expr` to do the check. I think it'd be good to leave that for a separate PR though.

File: crates/ra_hir/src/ty/infer.rs [hunk 8]
```diff
@@ -595,32 +647,41 @@ impl<'a, D: HirDatabase> InferenceContext<'a, D> {
...
                 let bound_ty = match mode {
-                    BindingAnnotation::Ref => Ty::Ref(inner_ty.clone().into(), Mutability::Shared),
-                    BindingAnnotation::RefMut => Ty::Ref(inner_ty.clone().into(), Mutability::Mut),
-                    BindingAnnotation::Mutable | BindingAnnotation::Unannotated => inner_ty.clone(),
+                    BindingMode::Ref(Mutability::Shared) => {
+                        Ty::Ref(inner_ty.clone().into(), Mutability::Shared)
+                    }
+                    BindingMode::Ref(Mutability::Mut) => {
+                        Ty::Ref(inner_ty.clone().into(), Mutability::Mut)
+                    }
```
Comment (flodiebold) [inline crates/ra_hir/src/ty/infer.rs]: these two could just be written as
```rust
BindingMode::Ref(m) => Ty::ref(inner_ty.clone().into(), m)
```
or am I missing something?

File: crates/ra_hir/src/ty/tests.rs (hunk 1 above)
Comment (flodiebold) [reply crates/ra_hir/src/ty/tests.rs]: maybe a separate test for that, plus a marker (`tested_by!`), would be nice :)

Comment (flodiebold) [reply crates/ra_hir/src/ty/tests.rs]: Also, in general, I'd prefer adding the new test cases in a separate test -- it's already not so easy what's what here

File: crates/ra_hir/src/ty/infer.rs (hunk 8 above)
Comment (mjkillough) [reply crates/ra_hir/src/ty/infer.rs]: You're right! That'd be much simpler.

File: crates/ra_hir/src/ty/tests.rs (hunk 1 above)
Comment (mjkillough) [reply crates/ra_hir/src/ty/tests.rs]: I hadn't come across `tested_by!`. Very neat! :slightly_smiling_face: 

I separated this out into two test-cases, as I think it makes the snapshots easier to read, and makes it clear what is being tested.

**Pull request merged**